File: objectiveai-py/objectiveai/push_utils.py
"""Shared helpers for push (streaming chunk accumulation)."""

from __future__ import annotations


def push_option_int(self_val: int | None, other_val: int | None) -> int | None:
    """Sum two optional ints (like Rust ``push_option_u64``)."""
    if self_val is not None and other_val is not None:
        return self_val + other_val
    if other_val is not None:
        return other_val
    return self_val


def push_option_string(self_val: str | None, other_val: str | None) -> str | None:
    """Concatenate two optional strings."""
    if self_val is not None and other_val is not None:
        return self_val + other_val
    if other_val is not None:
        return other_val
    return self_val


def push_option(self_val, other_val):
    """Merge two optional sub-objects by delegating to ``push()``.

    Both present → ``self_val.push(other_val)``, returns self_val.
    Only other → returns other_val (adopted).
    Only self / neither → returns self_val.
    """
    if self_val is not None and other_val is not None:
        self_val.push(other_val)
        return self_val
    if other_val is not None:
        return other_val
    return self_val


def push_by_index(self_list: list, other_list: list) -> None:
    """Merge *other_list* into *self_list* by ``index`` field.

    Items with a matching index are merged via ``push()``.
    New indices are appended.
    """
    from pydantic import RootModel

    index_map: dict[int, int] = {}
    for pos, item in enumerate(self_list):
        idx = _get_index(item)
        if idx is not None:
            index_map[idx] = pos

    for other_item in other_list:
        idx = _get_index(other_item)
        if idx is not None and idx in index_map:
            self_list[index_map[idx]].push(other_item)
        else:
            self_list.append(other_item)
            if idx is not None:
                index_map[idx] = len(self_list) - 1


def _get_index(item):
    """Extract an integer index from a model (BaseModel or RootModel)."""
    from pydantic import RootModel
    if isinstance(item, RootModel):
        inner = item.root
        if isinstance(inner, RootModel):
            inner = inner.root
        return getattr(inner, "index", None)
    return getattr(item, "index", None)

File: objectiveai-dotnet/ObjectiveAI/PushUtils.cs
namespace ObjectiveAI;

/// <summary>
/// Static push helper methods for streaming chunk accumulation.
/// Matches Go's push_utils.go and Python's push_utils.py.
/// </summary>
public static class PushUtils
{
    /// <summary>
    /// Merge items by matching a required index field.
    /// Items with matching index are merged via push; new items are appended.
    /// </summary>
    public static void PushByIndex<T>(List<T> self, List<T>? other, Func<T, long> getIndex, Action<T, T> push)
    {
        if (other == null || other.Count == 0) return;
        var indexMap = new Dictionary<long, int>();
        for (int i = 0; i < self.Count; i++)
            indexMap[getIndex(self[i])] = i;

        foreach (var item in other)
        {
            var idx = getIndex(item);
            if (indexMap.TryGetValue(idx, out var pos))
            {
                push(self[pos], item);
            }
            else
            {
                self.Add(item);
                indexMap[idx] = self.Count - 1;
            }
        }
    }

    /// <summary>
    /// Merge items by matching a nullable index field.
    /// Items with null index are always appended (never merged).
    /// </summary>
    public static void PushByNullableIndex<T>(List<T> self, List<T>? other, Func<T, long?> getIndex, Action<T, T> push)
    {
        if (other == null || other.Count == 0) return;
        var indexMap = new Dictionary<long, int>();
        for (int i = 0; i < self.Count; i++)
        {
            var idx = getIndex(self[i]);
            if (idx.HasValue)
                indexMap[idx.Value] = i;
        }

        foreach (var item in other)
        {
            var idx = getIndex(item);
            if (idx.HasValue && indexMap.TryGetValue(idx.Value, out var pos))
            {
                push(self[pos], item);
            }
            else
            {
                self.Add(item);
                if (idx.HasValue)
                    indexMap[idx.Value] = self.Count - 1;
            }
        }
    }

    /// <summary>Latest non-null value wins.</summary>
    public static T? PushReplace<T>(T? self, T? other) where T : class
        => other ?? self;

    /// <summary>Latest non-null value wins (value types).</summary>
    public static T? PushReplaceValue<T>(T? self, T? other) where T : struct
        => other ?? self;

    /// <summary>Concatenate optional strings.</summary>
    public static string? PushOptionString(string? self, string? other)
    {
        if (self != null && other != null) return self + other;
        return other ?? self;
    }

    /// <summary>Sum optional longs.</summary>
    public static long? PushOptionLong(long? self, long? other)
    {
        if (self.HasValue && other.HasValue) return self.Value + other.Value;
        return other ?? self;
    }

    /// <summary>Sum optional ulongs.</summary>
    public static ulong? PushOptionUlong(ulong? self, ulong? other)
    {
        if (self.HasValue && other.HasValue) return self.Value + other.Value;
        return other ?? self;
    }

    /// <summary>Sum optional doubles.</summary>
    public static double? PushOptionDouble(double? self, double? other)
    {
        if (self.HasValue && other.HasValue) return self.Value + other.Value;
        return other ?? self;
    }

    /// <summary>Monotonic boolean: once true, stays true.</summary>
    public static bool? PushLazySetTrue(bool? self, bool? other)
    {
        if (other.HasValue && other.Value) return true;
        return self;
    }

    /// <summary>
    /// Conditional merge of optional sub-objects.
    /// Both present → push; only other → adopt; otherwise no change.
    /// </summary>
    public static void PushOption<T>(ref T? self, T? other, Action<T, T> push) where T : class
    {
        if (self != null && other != null)
            push(self, other);
        else if (other != null)
            self = other;
    }
}
